extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java

```diff
@@ -165,26 +152,56 @@ public void logCleanup(BuildProducer<LogCleanupFilterBuildItem> logCleanupFilter
     void registerMetricsFromAnnotatedMethods(SmallRyeMetricsRecorder metrics,
             BeanContainerBuildItem beanContainerBuildItem,
             BeanArchiveIndexBuildItem beanArchiveIndex) {
-        JandexBeanInfoAdapter beanInfoAdapter = new JandexBeanInfoAdapter(beanArchiveIndex.getIndex());
-        JandexMemberInfoAdapter memberInfoAdapter = new JandexMemberInfoAdapter(beanArchiveIndex.getIndex());
-        for (ClassInfo clazz : beanArchiveIndex.getIndex().getKnownClasses()) {
-            // TODO: maybe leave out more cases where we can be sure that there are no metrics to be registered
-            // to speed up the startup
-            if (clazz.name().prefix().toString().startsWith("io.smallrye.metrics")) {
-                continue;
+        IndexView index = beanArchiveIndex.getIndex();
+        JandexBeanInfoAdapter beanInfoAdapter = new JandexBeanInfoAdapter(index);
+        JandexMemberInfoAdapter memberInfoAdapter = new JandexMemberInfoAdapter(index);
+
+        List<MethodInfo> collectedMetricsMethods = new ArrayList<>();
+        Map<DotName, ClassInfo> collectedMetricsClasses = new HashMap<>();
+
+        for (DotName metricAnnotation : METRICS_ANNOTATIONS) {
+            Collection<AnnotationInstance> metricAnnotationInstances = index.getAnnotations(metricAnnotation);
+            for (AnnotationInstance metricAnnotationInstance : metricAnnotationInstances) {
+                AnnotationTarget metricAnnotationTarget = metricAnnotationInstance.target();
+                switch (metricAnnotationTarget.kind()) {
+                    case METHOD: {
+                        MethodInfo method = metricAnnotationTarget.asMethod();
+                        if (!collectedMetricsMethods.contains(method)) {
```

Comment:
Maybe a `HashSet` would be better than a list?

Reply:
MethodInfo has no `hashCode` method

Reply:
Sure, thanks. I thought `MethodInfo` has no `hashCode` method just like `ClassInfo`. Will change.